docs/tm/udq_actionx/pyaction.tex

```diff
@@ -0,0 +1,137 @@
+the \actionx{} keyword, but instead of a \inlinecode{.DATA} formatted condition
+you are allowed to evaluate the condition with a general Python script. In
+principle the script can run arbitrary code, but due to the complexity of the
+\kw{SCHEDULE} datamodel the ``best'' way to actually change the course of the
+simulation is through the use of an additional dummy \actionx{} keyword.
+
+In order to enable the \pyaction{} keyword \flow{} must be compiled with the
+\path{cmake} switch \inlinecode{-DOPM\_ENABLE\_EMBEDDED\_PYTHON=ON}, the default is
+to build with \inlinecode{-DOPM\_ENABLE\_EMBEDDED\_PYTHON=OFF}.
+
+\section{Python - wrapping and embedding}
+Python is present in the \flow{} codebase in two different ways. For many of
+the classes in the \flow{} codebase - in particular in opm-common, there are
+\emph{Python wrappers} available. That means that you can invoke the C++
+functionality in \flow{} classes from Python - e.g. this Python script can be used to
+load a deck and print all the keywords:
+
+\begin{code}
+import sys
+from opm.io.parser import Parser
+
+input_file = sys.argv[1]
+parser = Parser()
+
+deck = parser.parse_file(input_file)
+for kw in deck:
+    print(kw.name)
+\end{code}
+ 
+When used this way the Python interpreter is the main program running, and the
+\flow{} classes like \inlinecode{Opm::Parser} are loaded to extend the Python
+interpreter. This can also be flipped around, the Python interpreter can be
+\emph{embedded} in the \flow{} executable. When Python is embedded, \flow{} is
+the main program running, and with help of the embedded interpreter the \flow{}
+program can be extended with Python plugins. The \pyaction{} keyword can be
+perceived as a Python plugin. To really interact with the state of the \flow{}
+simulation the plugin needs to utilize the functionality which wraps the C++
+functionality.
+
+\section{The \pyaction{} keyword}
+The \pyaction{} keyword is in the \kw{SCHEDULE} section like \actionx{}. The
+first record is the name of the action and a string identifier for how many
+times the action should run, then there is a path to Python module:
+
+\begin{deck}
+PYACTION
+  PYTEST 'FIRST_TRUE' /
+  'pytest.py' /
+\end{deck}
+
+This keyword defines a \pyaction{} called \kw{PYTEST} which will run at the end
+of every timestep until the first time a \inlinecode{true} value is returned. In
+addition to \kw{FIRST\_TRUE} you can choose \kw{SINGLE} to run exactly once and
+\kw{UNLIMITED} to continue running at the end of every timestep for the entire
+simulation. The second record is the path to a file with Python code which will
+run when this \pyaction{} is invoked. The path to the module will be interpreted
+relative to the location of the \path{.DATA} file.
+
+The python module can be quite arbitrary, but it must contain a function
+\inlinecode{run} with the correct signature:
+\begin{code}
+def run(ecl_state, schedule, report_step, summary_state, actionx_callback):
```

Comment:
I think we should explain what the parameters are. At least the last one is not obvious,

Reply:
That is done in some detail in section 3.2.1